src/client-side-encryption/client_encryption.ts

```diff
@@ -541,13 +559,20 @@ export class ClientEncryption {
       }
     } = options;
 
+    const timeoutContext = this._timeoutMS
+      ? new CSOTTimeoutContext({
+          timeoutMS: this._timeoutMS,
+          serverSelectionTimeoutMS: this._client.options.serverSelectionTimeoutMS
+        })
+      : undefined;
+
     if (Array.isArray(encryptedFields.fields)) {
       const createDataKeyPromises = encryptedFields.fields.map(async field =>
         field == null || typeof field !== 'object' || field.keyId != null
           ? field
           : {
               ...field,
-              keyId: await this.createDataKey(provider, { masterKey })
+              keyId: await this.createDataKey(provider, { masterKey, timeoutContext })
```

Comment:
Bailey pointed out to me that this may be an issue since we're using the same timeout context concurrently, which is not something I think our timeoutContext can do. Something like this to `CSOTTimeoutContext` in  src/timeout.ts:

```ts
  clone(): CSOTTimeoutContext {
    const timeoutContext = new CSOTTimeoutContext({
      timeoutMS: this.timeoutMS,
      serverSelectionTimeoutMS: this.serverSelectionTimeoutMS
    });
    timeoutContext.start = this.start;

    return timeoutContext;
  }

```

And setting the timeoutContext being used in createDataKey here to `timeoutContext.clone()`

Reply:
I see, I added in the .clone() function and call so there's no concurrent timeoutContexts being used!